src/main/java/com/google/cloud/tools/project/AppYaml.java

```diff
@@ -16,54 +16,113 @@
-    try (InputStream in = Files.newInputStream(appYaml)) {
-      Object loaded = new Yaml().load(in);
-      if (loaded == null) {
-        yamlMap = Collections.emptyMap();
-      } else {
-        yamlMap = (Map<String, ?>) loaded;
-      }
-    }
+  @SuppressWarnings("unchecked")
+  public static AppYaml parse(InputStream input) {
+    // our needs are simple so just load using primitive objects
+    Yaml yaml = new Yaml(new SafeConstructor());
+    Map<String, ?> contents = (Map<String, ?>) yaml.load(input);
+    return new AppYaml(contents);
+  }
+
+  private AppYaml(Map<String, ?> yamlMap) {
+    this.yamlMap = yamlMap == null ? Collections.emptyMap() : yamlMap;
   }
 
   /**
-   * Returns "runtime" value if it is a String type. {@code null} if it is not a String or the
-   * "runtime" key is not present.
+   * Return the content of the {@code environment} field, which defines this app's required App
+   * Engine environment.
    */
   @Nullable
+  public String getEnvironmentType() {
+    return getString(ENVIRONMENT_TYPE_KEY);
+  }
+
+  /** Return the content of the {@code runtime} field, which defines this app's expected runtime. */
+  @Nullable
   public String getRuntime() {
-    Object result = yamlMap.get(RUNTIME_KEY);
-    if (result instanceof String) {
-      return (String) result;
+    return getString(RUNTIME_KEY);
+  }
+
+  /**
+   * Return the content of the {@code application} field, which identifies the App Engine API
+   * version used by this app.
+   */
+  @Nullable
+  public String getApiVersion() {
+    return getString(API_VERSION_KEY);
+  }
+
+  /** Return the content of the {@code application} field, which identifies the Project ID. */
+  @Nullable
+  public String getProjectId() {
+    return getString(APPLICATION_KEY);
+  }
+
+  /** Return the content of the {@code version} field, which identifies the Project version. */
+  @Nullable
+  public String getProjectVersion() {
+    return getString(VERSION_KEY);
+  }
+
+  /**
+   * Return the content of the {@code service} field or the now-deprecated {@code module} field,
+   * which identifies the Service ID.
+   */
+  @Nullable
+  public String getServiceId() {
+    String serviceId = getString(SERVICE_KEY);
+    if (serviceId == null) {
+      serviceId = getString(MODULE_KEY);
     }
+    return serviceId;
+  }
+
+  /** Return the content of the {@code env_variables} field, which defines environment variables. */
+  @Nullable
+  public Map<String, ?> getEnvironmentVariables() {
+    return getStringMap(ENVIRONMENT_VARIABLES_KEY);
+  }
 
-    return null;
+  @Nullable
+  private String getString(String key) {
+    Object value = yamlMap.get(key);
+    return value instanceof String ? (String) value : null;
+  }
+
+  @Nullable
+  @SuppressWarnings("unchecked")
+  private Map<String, ?> getStringMap(String key) {
+    Object value = yamlMap.get(key);
+    return value instanceof Map<?, ?> ? (Map<String, ?>) value : null;
```

Comment:
Probably `instanceof Map<String, ?>`.

Reply:
That causes an error in Eclipse since you can't actually check the generic types.